ports/stm/supervisor/internal_flash.c

```diff
@@ -101,9 +159,82 @@ uint32_t supervisor_flash_get_block_count(void) {
 }
 
 void supervisor_flash_flush(void) {
+    if (_cache_flash_addr == NO_CACHE) return;
+
+    #if defined(STM32H7)
+    __HAL_FLASH_CLEAR_FLAG(FLASH_FLAG_ALL_ERRORS_BANK1 | FLASH_FLAG_ALL_ERRORS_BANK2);
+    #else
+    __HAL_FLASH_CLEAR_FLAG(FLASH_FLAG_EOP | FLASH_FLAG_OPERR | FLASH_FLAG_WRPERR |
+                           FLASH_FLAG_PGAERR | FLASH_FLAG_PGPERR | FLASH_FLAG_PGSERR);
+    #endif
+
+    // set up for erase
+    FLASH_EraseInitTypeDef EraseInitStruct;
+    EraseInitStruct.TypeErase = TYPEERASE_SECTORS;
+    EraseInitStruct.VoltageRange = VOLTAGE_RANGE_3; // voltage range needs to be 2.7V to 3.6V
+    // get the sector information
+    uint32_t sector_size;
+    uint32_t sector_start_addr;
+    #if defined(STM32H7)
+    EraseInitStruct.Banks = get_bank(_cache_flash_addr);
+    #endif
+    EraseInitStruct.Sector = flash_get_sector_info(_cache_flash_addr, &sector_start_addr, &sector_size);
+    EraseInitStruct.NbSectors = 1;
+    if (sector_size > sizeof(_flash_cache)) {
+        reset_into_safe_mode(FLASH_WRITE_FAIL);
+    }
+
+    // Skip if data is the same
+    if (memcmp(_flash_cache, (void *)_cache_flash_addr, sector_size) != 0) {
+        // unlock flash
+        HAL_FLASH_Unlock();
+
+        // erase the sector
+        uint32_t SectorError = 0;
+        if (HAL_FLASHEx_Erase(&EraseInitStruct, &SectorError) != HAL_OK) {
+            // error occurred during sector erase
+            HAL_FLASH_Lock(); // lock the flash
+            reset_into_safe_mode(FLASH_WRITE_FAIL);
+        }
+
+        uint32_t * cache_addr = (uint32_t*)_flash_cache;
+
+        #if defined(STM32H7)
+        for (uint32_t i = 0; i < (sector_size / 32); i++) {
+            // Note that the STM32H7 HAL interface differs by taking an address, not 64 bit data
+            // This is because ST's code is written by a large room of chimpanzees
```

Comment:
```suggestion
```

Be kind and please review [the Code of Conduct](https://github.com/adafruit/Adafruit_Community_Code_of_Conduct/blob/master/code-of-conduct.md). Comments like this are unacceptable.